vector/src/main/java/im/vector/app/features/lifecycle/VectorActivityLifecycleCallbacks.kt

```diff
@@ -16,31 +16,122 @@
-    }
+    override fun onActivityStarted(activity: Activity) {}
 
-    override fun onActivityDestroyed(activity: Activity) {
-    }
+    override fun onActivityDestroyed(activity: Activity) {}
 
-    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
-    }
+    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
 
-    override fun onActivityStopped(activity: Activity) {
-    }
+    override fun onActivityStopped(activity: Activity) {}
 
     override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
+        // restart the app if the task contains an unknown activity
+        coroutineScope.launch {
+            if (isTaskCorrupted(activity)) {
+                Timber.e("Application is potentially corrupted by an unknown activity")
+                MainActivity.restartApp(activity, MainActivityArgs())
+                return@launch
+            }
+        }
     }
+
+    /**
+     * Check if all activities running on the task with package name affinity are safe.
+     *
+     * @return true if an app task is corrupted by a potentially malicious activity
+     */
+    @SuppressLint("NewApi")
+    @Suppress("DEPRECATION")
+    private suspend fun isTaskCorrupted(activity: Activity): Boolean = withContext(Dispatchers.Default) {
+        val context = activity.applicationContext
+        val packageManager: PackageManager = context.packageManager
+
+        try {
+            // Get all activities from app manifest
+            if (activitiesInfo.isEmpty()) {
+                activitiesInfo = packageManager.getPackageInfo(context.packageName, PackageManager.GET_ACTIVITIES).activities
+            }
+
+            // Get all running activities on app task
+            // and compare to activities declared in manifest
+            val manager = context.getSystemService<ActivityManager>() ?: return@withContext false
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                // Android lint may return an error on topActivity field.
+                // This field was added in ActivityManager.RecentTaskInfo class since Android M (API level 23)
+                // and it is inherited from TaskInfo since Android Q (API level 29).
+                // API 23 changes : https://developer.android.com/sdk/api_diff/23/changes/android.app.ActivityManager.RecentTaskInfo
+                // API 29 changes : https://developer.android.com/sdk/api_diff/29/changes/android.app.ActivityManager.RecentTaskInfo
+                try {
+                    manager.appTasks.any { appTask ->
+                        appTask.taskInfo.topActivity?.let { isPotentialMaliciousActivity(it) } ?: false
+                    }
+                } catch (e: IllegalArgumentException) {
+                    // The task was not found. We can ignore it.
+                    Timber.e("The task was not found: ${e.localizedMessage}")
+                    false
+                }
```

Comment:
As we are already in a try block, would it be better to catch this error in the other try?